src/metatensor/models/utils/data/readers/targets/ase.py

```diff
@@ -1,43 +1,157 @@
+        ),
+    )
 
-        n_structures = len(values)
+    values = torch.concatenate(values_raw, dim=0)
+    assert values.shape[1] != (3,)
 
-        block = TensorBlock(
-            values=torch.tensor(values).reshape(-1, 1),
-            samples=Labels(["structure"], torch.arange(n_structures).reshape(-1, 1)),
-            components=[],
-            properties=Labels.single(),
-        )
+    block = TensorBlock(
+        values=values.reshape(-1, 3, 1),
+        samples=samples,
+        components=[Labels(["direction"], torch.arange(3).reshape(-1, 1))],
+        properties=Labels.single(),
+    )
+
+    return block
 
-        target_dictionary[target_value] = TensorMap(
-            keys=Labels(["lambda", "sigma"], torch.tensor([(0, 1)])), blocks=[block]
+
+def read_virial_ase(
+    filename: str,
+    key: str,
+):
+    """Store virial information in :class:`metatensor.TensorBlock` which can be used as
+    ``displacement`` gradients
+
+    :param filename: name of the file to read
+    :param key: target value key name to be parsed from the file
+
+    :returns:
+        TensorMap containing the given information
+    """
+    return _read_virial_stress_ase(filename=filename, key=key, is_virial=True)
+
+
+def read_stress_ase(
+    filename: str,
+    key: str,
+):
+    """Store stress information in :class:`metatensor.TensorBlock` which can be used as
+    ``displacement`` gradients
+
+    :param filename: name of the file to read
+    :param key: target value key name to be parsed from the file
+
+    :returns:
+        TensorMap containing the given information
+    """
+    return _read_virial_stress_ase(filename=filename, key=key, is_virial=False)
+
+
+def _read_virial_stress_ase(
+    filename: str,
+    key: str,
+    is_virial: bool = True,
+) -> TensorBlock:
+    """Store stress or virial information in :class:`metatensor.TensorBlock` which can
+    be used as ``displacement`` gradients
+
+    :param filename: name of the file to read
+    :param key: target value key name to be parsed from the file
+    :param is_virial: if target values are stored as stress or virials.
+
+    :returns:
+        TensorMap containing the given information
+    """
+    frames = ase.io.read(filename, ":")
+    n_structures = len(frames)
+    values = torch.tensor([f.info[key].tolist() for f in frames])
+
+    if values.shape[1:] == (9,):
+        warnings.warn(
+            "Found 9-long numerical vector for the stress/virial. Assume a row major "
+            "format for the conversion into a 3x3 matrix.",
+            stacklevel=2,
         )
+        values = values.reshape(n_structures, 3, 3)
+    elif values.shape[1:] != (3, 3):
+        raise ValueError(
+            f"stress/virial must be a 3 x 3 matrix but has shape {values.shape}"
+        )
+
+    if is_virial:
+        values *= -1
+    else:  # is stress
+        values *= torch.tensor([f.cell.tolist() for f in frames])
```

Comment:
Here I think `values` should multiplied by the volume of the cell (ASE might have a function or property for that). I'm not sure what this operation is doing. It would also be great to raise a good error if there is no cell.

Reply:
Ohh yes you are completely, right!